### PR TITLE
Ratis Server Implementation with Resource Manager

### PR DESCRIPTION
**Pull request by kaur-h**
Description - Introducing the raft consensus algorithm among resource managers using the Apache Ratis library. Each resource manager integrates a raft server which allows one of the resource managers to be chosen as the leader with which the coordinator can communicate with. 

Test plan - Created a unit test that runs multiple resource managers and builds a client. The client is used to check that one of the resource managers is elected as a leader in the leader election process of the raft algorithm. 

```
== RELEASE NOTES ==

General Changes
* Add raft server to resource manager with disaggregated coordinators using Apache Ratis. Raft server enabled in resource manager using the ``isEnabled`` property in Config file 
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    private String raftGroupId;
+    private File storageDir;
+    private int raftPort;
+
+    public String getId()
+    {
+        return id;
+    }
+
+    @Config("ratis-id")
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java:19]: call it raft.id

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    public RaftConfig setStorageDir(String storageDir) {
+        this.storageDir = new File(storageDir + "/" + id);
+        return this;
+    }
+
+    public String getRaftGroupId() {
+        return raftGroupId;
+    }
+
+    @Config("ratis.groupId")
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java:38]: raft.groupId

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    public RaftConfig setRaftGroupId(String groupId) {
+        this.raftGroupId = groupId;
+        return this;
+    }
+
+    public int getRaftPort() {
+        return raftPort;
+    }
+
+    @Config("ratis.raftPort")
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java:48]: raft.port

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import com.facebook.airlift.configuration.Config;
+
+import java.io.File;
+
+public class RaftConfig
+{
+    private String id;
+    private String raftGroupId;
+    private File storageDir;
+    private int raftPort;
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java:12]: call it just `port`

File: presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java [hunk 5]
```diff
@@ -98,9 +100,10 @@ public void getClusterStats(
             @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
             @Context UriInfo uriInfo,
             @Context HttpServletRequest servletRequest,
-            @Suspended AsyncResponse asyncResponse)
+            @Suspended AsyncResponse asyncResponse,
+            @QueryParam("includeLocalInfoOnly") @DefaultValue("false") boolean includeLocalInfoOnly)
     {
-        if (resourceManagerEnabled) {
+        if (resourceManagerEnabled && !includeLocalInfoOnly) {
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/server/ClusterStatsResource.java:106]: Looks like this are not related changes. Please remove them.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 6]
```diff
@@ -609,11 +610,19 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
...
+            raftPortProperty = String.valueOf(raftConfig.getRaftPort());
+        }
+
         discoveryBinder(binder).bindHttpAnnouncement("presto")
                 .addProperty("node_version", nodeVersion.toString())
                 .addProperty("coordinator", String.valueOf(serverConfig.isCoordinator()))
                 .addProperty("resource_manager", String.valueOf(serverConfig.isResourceManager()))
-                .addProperty("connectorIds", nullToEmpty(serverConfig.getDataSources()));
+                .addProperty("connectorIds", nullToEmpty(serverConfig.getDataSources()))
+                .addProperty("raftPort", raftPortProperty);
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:625]: can we add this property only for resource manager?

File: presto-tests/src/test/java/com/facebook/presto/resourcemanager/TestDistributedClusterStatsResource.java [hunk 7]
```diff
@@ -62,13 +62,14 @@
...
     private TestingPrestoServer coordinator2;
+    private DistributedQueryRunner runner;
 
     @BeforeClass
     public void setup()
             throws Exception
     {
         client = new JettyHttpClient();
-        DistributedQueryRunner runner = createQueryRunner(
+        runner = createQueryRunner(
```
Comment (swapsmagic) [inline presto-tests/src/test/java/com/facebook/presto/resourcemanager/TestDistributedClusterStatsResource.java:72]: Looks like not related changes. Let's remove them from here.

**Review by swapsmagic — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java [hunk 8]
```diff
@@ -507,6 +509,19 @@ private static OptionalInt getThriftServerPort(ServiceDescriptor descriptor)
...
     }
 
+    private static OptionalInt getRaftPort(ServiceDescriptor descriptor)
+    {
+        String port = descriptor.getProperties().get("raftPort");
+        if (port != "") {
+            try {
+                return OptionalInt.of(Integer.parseInt(port));
+            }
+            catch (IllegalArgumentException ignored) {
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:519]: Let's log the error message here so it won't get unnoticed.

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    public int getPort() {
+        return port;
+    }
+
+    @Config("raft.port")
+    public RaftConfig setPort(int port) {
+        this.port = port;
+        return this;
+    }
+}
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RaftConfig.java:53]: new line needed

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+       }).toArray(RaftPeer[]::new);
+    }
+
+    @PostConstruct
+    public void startServer()
+            throws Exception
+    {
+        run();
+    }
+    public void run() throws Exception
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java:77]: add space between methods.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 11]
```diff
@@ -265,7 +266,7 @@ protected void setup(Binder binder)
     {
         ServerConfig serverConfig = buildConfigObject(ServerConfig.class);
 
-        if (serverConfig.isResourceManager()) {
+        if (serverConfig.isResourceManager ()) {
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:269]: remove the extra space.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 12]
```diff
@@ -609,11 +610,23 @@ public ListeningExecutorService createResourceManagerExecutor(ResourceManagerCon
...
+                    .addProperty("resource_manager", String.valueOf(serverConfig.isResourceManager()))
+                    .addProperty("connectorIds", nullToEmpty(serverConfig.getDataSources()))
+                    .addProperty("raftPort", String.valueOf(raftConfig.getPort()));
+        }
+        else{
+            discoveryBinder(binder).bindHttpAnnouncement("presto")
+                    .addProperty("node_version", nodeVersion.toString())
+                    .addProperty("coordinator", String.valueOf(serverConfig.isCoordinator()))
+                    .addProperty("resource_manager", String.valueOf(serverConfig.isResourceManager()))
+                    .addProperty("connectorIds", nullToEmpty(serverConfig.getDataSources()));
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java:628]: let's merge this and add raft property only for resource manager.

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 13]
```diff
@@ -198,8 +202,16 @@ private DistributedQueryRunner(
             extraCoordinatorProperties.putAll(coordinatorProperties);
 
             if (resourceManagerEnabled) {
-                resourceManager = Optional.of(closer.register(createTestingPrestoServer(discoveryUrl, true, true, false, resourceManagerProperties, parserOptions, environment, baseDataDir, extraModules)));
-                servers.add(resourceManager.get());
+//                resourceManager = Optional.of(closer.register(createTestingPrestoServer(discoveryUrl, true, true, false, resourceManagerProperties, parserOptions, environment, baseDataDir, extraModules)));
+//                servers.add(resourceManager.get());
```
Comment (swapsmagic) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java:206]: remove the commented code if not needed.

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 14]
```diff
@@ -150,7 +151,8 @@ private DistributedQueryRunner(
             String environment,
             Optional<Path> baseDataDir,
             Optional<BiFunction<Integer, URI, Process>> externalWorkerLauncher,
-            List<Module> extraModules)
+            List<Module> extraModules,
+            int resourceManagerCount)
```
Comment (swapsmagic) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java:155]: let's add `resourceManagerCount` closer to the flag `resourceManagerEnabled`

**Review by swapsmagic — CHANGES_REQUESTED** [review]

File: presto-main/pom.xml [hunk 15]
```diff
@@ -435,10 +435,71 @@
...
+            <optional>true</optional>
+        </dependency>
     </dependencies>
 
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.basepom.maven</groupId>
+                <artifactId>duplicate-finder-maven-plugin</artifactId>
+                <version>1.2.1</version>
```
Comment (swapsmagic) [inline presto-main/pom.xml:469]: let's define version above and reuse that like we do it for other artifacts.

File: presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java [hunk 16]
```diff
@@ -507,6 +509,20 @@ private static OptionalInt getThriftServerPort(ServiceDescriptor descriptor)
...
 
+    private static OptionalInt getRaftPort(ServiceDescriptor descriptor)
+    {
+        String port = descriptor.getProperties().get("raftPort");
+        if (port != null) {
+            try {
+                return OptionalInt.of(Integer.parseInt(port));
+            }
+            catch (IllegalArgumentException exception) {
+                log.error(exception, "Error getting raft port");
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java:520]: Let's log the actual value of raftPort that we get from the service descriptor.

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        return resourceManagers.stream().map(resourceManager -> {
+            RaftPeer.Builder builder = RaftPeer.newBuilder();
+            builder.setId(RaftPeerId.valueOf(resourceManager.getNodeIdentifier()))
+                   .setAddress(resourceManager.getHost() + ":" + resourceManager.getRaftPort().getAsInt());
+            return builder.build();
+        }).toArray(RaftPeer[]::new);
+    }
+
+    @PostConstruct
+    public void startServer()
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java:78]: let's name it: `start`

File: presto-tests/pom.xml [hunk 18]
```diff
@@ -280,10 +280,80 @@
...
+        </dependency>
+        
     </dependencies>
 
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.basepom.maven</groupId>
+                <artifactId>duplicate-finder-maven-plugin</artifactId>
+                <version>1.2.1</version>
```
Comment (swapsmagic) [inline presto-tests/pom.xml:323]: is this required?

**Comment by ajaygeorge** [comment]
@kaur-h Let us squash the [Dependency Fix](https://github.com/prestodb/presto/pull/17853/commits/0c8110a4979e1fece9be281200f9cd8ac0b94a66) commit.

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        RaftPeerId peerId = RaftPeerId.valueOf(id);
+        RaftProperties properties = new RaftProperties();
+
+        GrpcConfigKeys.Server.setPort(properties, port);
+
+        File storage = new File(storageDir + "/" + id);
+        RaftServerConfigKeys.setStorageDir(properties, Collections.singletonList(storage));
+
+        final RaftGroup raftGroup = RaftGroup.valueOf(RaftGroupId.valueOf(UUID.nameUUIDFromBytes(groupId.getBytes())), getPeers());
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java:94]: Can you help me understand the logic around `UUID.nameUUIDFromBytes` .
Since groupId is something that is set from static config, how would it be a UUID?

**Review by ajaygeorge — COMMENTED** [review]
Added some comments.

**Review by tdcmeehan — COMMENTED** [review]
This looks good, but I'm a little concerned about the heavy dependencies we'll be bringing into presto main.  Any way to isolate them to just the resource managers?

File: presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java (hunk 19 above)
Comment (kaur-h) [reply presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java:94]: RaftGroupId's are of type UUID so even though the groupId is set as a string from the config it needs to be converted to a UUID in order to build the RaftGroup.

Comment (ajaygeorge) [reply presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java:83]: Is the expectation that in the static config we set a UUID string and then convert it back to UUID object using `UUID.nameUUIDFromBytes` . If that is the case wondering why it should be mentioned in config. Can't we generate a group ID as a `UUID.randomUUID` in code. In other words, should the group ID be even part of the config.

Comment (kaur-h) [reply presto-main/src/main/java/com/facebook/presto/resourcemanager/RatisServer.java:83]: The groupId would also be needed when we build the RaftClient for the coordinator as the servers and client belong to the same RaftGroup.

**Comment by kaur-h** [comment]
> This looks good, but I'm a little concerned about the heavy dependencies we'll be bringing into presto main. Any way to isolate them to just the resource managers?

The only way to isolate them to just the resource manager would be to move the code into a different package so that the dependencies will be removed from presto main and into another package.

**Review by ajaygeorge — APPROVED** [review]
LGTM
